iceoryx_hoofs/posix/filesystem/source/file.cpp

```diff
@@ -0,0 +1,438 @@
+    case EROFS:
+        IOX_LOG(ERROR) << "Unable to remove file since it resides on a read-only file system.";
+        return iox::error<FileRemoveError>(FileRemoveError::ReadOnlyFilesystem);
+    default:
+        IOX_LOG(ERROR) << "Unable to remove file since an unknown error occurred.";
+        return iox::error<FileRemoveError>(FileRemoveError::UnknownError);
+    }
+}
+
+expected<FileOffsetError> File::set_offset(const uint64_t offset) const noexcept
+{
+    auto result = posix::posixCall(iox_lseek)(m_file_descriptor, static_cast<iox_off_t>(offset), IOX_SEEK_SET)
+                      .failureReturnValue(-1)
+                      .evaluate();
+
+    if (!result.has_error())
+    {
+        if (result->value == static_cast<iox_off_t>(offset))
+        {
+            return iox::success<>();
+        }
+
+        IOX_LOG(ERROR) << "Unable to set file offset position since it set to the wrong offset position.";
+        return iox::error<FileOffsetError>(FileOffsetError::OffsetAtWrongPosition);
+    }
+
+
+    switch (result.get_error().errnum)
+    {
+    case EINVAL:
+        IOX_FALLTHROUGH;
+    case ENXIO:
+        IOX_LOG(ERROR) << "Unable to set file offset position since it beyond the file limits.";
+        return iox::error<FileOffsetError>(FileOffsetError::OffsetBeyondFileLimits);
+    case EOVERFLOW:
+        IOX_LOG(ERROR)
+            << "Unable to set file offset position since the file is too large and the offset would overflow.";
+        return iox::error<FileOffsetError>(FileOffsetError::FileOffsetOverflow);
+    case EPIPE:
+        IOX_LOG(ERROR) << "Unable to set file offset position since seeking is not supported by the file type.";
+        return iox::error<FileOffsetError>(FileOffsetError::SeekingNotSupportedByFileType);
+    default:
+        IOX_LOG(ERROR) << "Unable to remove file since an unknown error occurred.";
+        return iox::error<FileOffsetError>(FileOffsetError::UnknownError);
+    }
+}
+
+expected<uint64_t, FileReadError> File::read(uint8_t* const buffer, const uint64_t buffer_len) const noexcept
+{
+    return read_at(0, buffer, buffer_len);
+}
+
+expected<uint64_t, FileReadError>
+File::read_at(const uint64_t offset, uint8_t* const buffer, const uint64_t buffer_len) const noexcept
+{
+    if (m_access_mode == posix::AccessMode::WRITE_ONLY)
+    {
+        IOX_LOG(ERROR) << "Unable to read from file since it is opened for writing only.";
+        return iox::error<FileReadError>(FileReadError::NotOpenedForReading);
+    }
+
+    if (set_offset(offset).has_error())
+    {
+        IOX_LOG(ERROR) << "Unable to read from file since the offset could not be set.";
+        return iox::error<FileReadError>(FileReadError::OffsetFailure);
+    }
+
+    auto result = posix::posixCall(iox_read)(m_file_descriptor, buffer, buffer_len).failureReturnValue(-1).evaluate();
+
+    if (!result.has_error())
+    {
+        return iox::success<uint64_t>(static_cast<uint64_t>(result->value));
```

Comment:
We desperately need automatic conversion from `expected<T, E1>` to `expected<T, E2>`